### PR TITLE
Make resource_id / nft_id required

### LINKED ISSUE #169: Resources should have assignable (not auto-incrementing) IDs, at least optionally
Similar to Collection/Nfts discussed here https://github.com/rmrk-team/rmrk-substrate/issues/83

Front-end implementations seem to be urging for the option of assigning your own ResourceID number, which would allow for "knowing" the ID ahead of time, instead of relying on the event revealing the ID assigned.

Opening this to document the conversation/discussion.

### PR DESCRIPTION
**Pull request by ilionic**
Fixes #169 
- [x] make resource_id required ( remove auto increment )
- [x]  make nft_id required ( remove auto increment )
- [x] make Basic resource metadata required
- [x] integration tests

### CODE REVIEW

File: pallets/rmrk-core/src/functions.rs [hunk 1]
```diff
@@ -330,19 +337,20 @@ where
...
 		collection_id: CollectionId,
 		royalty_recipient: Option<T::AccountId>,
 		royalty_amount: Option<Permill>,
 		metadata: StringLimitOf<T>,
 		transferable: bool,
-		resources: Option<BoundedResourceTypeOf<T>>,
+		resources: Option<BoundedResourceInfoTypeOf<T>>,
 	) -> sp_std::result::Result<(CollectionId, NftId), DispatchError> {
-		let nft_id = Self::get_next_nft_id(collection_id)?;
+		ensure!(Nfts::<T>::get(collection_id, nft_id).is_none(), Error::<T>::NftAlreadyExists);
```
Comment (HashWarlock) [inline pallets/rmrk-core/src/functions.rs:348]: Use `nft_exists` instead to check. We can likely refactor the function to return an Error.

Comment (ilionic) [reply pallets/rmrk-core/src/functions.rs:346]: Thanks! Good point

Comment (ilionic) [reply pallets/rmrk-core/src/functions.rs]: Updated, won't change it for now to return an error as it's used for other purposes for equippables

File: traits/src/resource.rs [hunk 2]
```diff
@@ -110,6 +107,20 @@ pub struct ResourceInfo<BoundedString, BoundedParts> {
...
+#[derive(Encode, Decode, Eq, PartialEq, Clone, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[cfg_attr(feature = "std", derive(Serialize))]
+#[cfg_attr(
+	feature = "std",
+	serde(bound = r#"
+			BoundedString: AsRef<[u8]>,
+			BoundedParts: AsRef<[PartId]>
+		"#)
+)]
+pub struct ResourceInfoMin<BoundedString, BoundedParts> {
```
Comment (HashWarlock) [inline traits/src/resource.rs:119]: What does `Min` in `ResourceInfoMin` stands for?

File: pallets/rmrk-core/src/functions.rs [hunk 3]
```diff
@@ -405,19 +418,20 @@ where
...
 	) -> sp_std::result::Result<(CollectionId, NftId), DispatchError> {
-		let nft_id = Self::get_next_nft_id(collection_id)?;
+		ensure!(!Self::nft_exists((collection_id, nft_id)), Error::<T>::NftAlreadyExists);
 		let collection = Self::collections(collection_id).ok_or(Error::<T>::CollectionUnknown)?;
 
-		// Prevent minting when next NFT id is greater than the collection max.
+		// Prevent minting when nfts_count is greater than the collection max.
 		if let Some(max) = collection.max {
-			ensure!(nft_id < max, Error::<T>::CollectionFullOrLocked);
+			ensure!(collection.nfts_count < max, Error::<T>::CollectionFullOrLocked);
```
Comment (HashWarlock) [inline pallets/rmrk-core/src/functions.rs:434]: Should we add a helper function that gets the Collection's NFT count?

File: traits/src/resource.rs (hunk 2 above)
Comment (ilionic) [reply traits/src/resource.rs:119]: Open for better naming suggestions.
 ResourceInfo - contains extra props, ResourceInfoMin - 'minimal', has only `id`. 
Another option was  `ResourceWithid`. wdyt?

Comment (HashWarlock) [reply traits/src/resource.rs:119]: What is the case for using `ResourceInfo` vs `ResourceInfoMin`?

Comment (ilionic) [reply traits/src/resource.rs:119]: Use case - adding resources with explicit id during minting. 
https://github.com/rmrk-team/rmrk-substrate/blob/d5c290d3053ebdc289990444b42aa40f3baf8daa/pallets/rmrk-core/src/functions.rs#L344

image

So it's similar to https://github.com/rmrk-team/rmrk-substrate/blob/d5c290d3053ebdc289990444b42aa40f3baf8daa/traits/src/resource.rs#L96-L108

Just without `pending_removal` and `pending` fields.

Comment (HashWarlock) [reply traits/src/resource.rs:119]: I think the old naming `ResourceInfoWithId` was reasonable, but the proposed `ResourceInfoMin` should work. So whichever one is okay with me.

File: pallets/rmrk-core/src/functions.rs (hunk 3 above)
Comment (HashWarlock) [reply pallets/rmrk-core/src/functions.rs:434]:
```
pub fn get_collection_nfts_count(collection_info: CollectionInfoOf<T>) -> u32 {
		collection_info.nfts_count
}
```

**Comment by ilionic** [comment]
> Should we think about adding `BoundedVec` for `ResourceId` and `NftId` now? This makes more sense now that we are not auto-incrementing the `NftId` and `ResourceId`. Do you think we should also propose this for the Uniques v2 implementation as well?

Do you mean using a 'stringy' version instead of u32? There was another point in favour of u32 - to keep it compatible with EVM.

File: pallets/rmrk-core/src/functions.rs (hunk 3 above)
Comment (ilionic) [reply pallets/rmrk-core/src/functions.rs:434]: I am not sure actually, seems a bit unneeded abstraction of just accessing property? But happy to update if you think worth pursuing. ( Usually I would extract a function with more logic )

**Review by HashWarlock — APPROVED** [review]
LGTM

**Pull request merged**